tensorboard/webapp/widgets/line_chart_v2/lib/coordinator.ts

```diff
@@ -0,0 +1,110 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+import {createScale, Scale} from './scale';
+import {Rect, ScaleType} from './types';
+import {convertRectToExtent} from './utils';
+
+/**
+ * A stateful convenient utility around scale for converting coordinate systems.
+ *
+ * Definitions.
+ *
+ * Example for better illustration: we are viewing a diagonal line that goes from
+ * <0, 0> -> <1, 2> onto a canvas with size <100, 200>.
+ *
+ * - data coordinate: coordinate in raw data space. For example above, you would have a
+ *     line by connecting two points at <0, 0> and <1, 2>.
+ * - ui coordinate: coordinate of a data in pixel/view-space. For example above, a data at
+ *     <0.5, 0.5> will be on <50, 100> in UI coordinates.
+ * - internal coordinate: in case like webgl, you can use an internal static
+ *     coordinate system separate from the ui coordinate.
+ * - view box: a rect in data coordinate that describes what should be visible
+ *     on the screen.
+ */
+export class Coordinator {
+  protected xScale: Scale = createScale(ScaleType.LINEAR);
+  protected yScale: Scale = createScale(ScaleType.LINEAR);
+  protected domContainerRect: Rect = {
+    x: 0,
+    width: 1,
+    y: 0,
+    height: 1,
+  };
+
+  protected lastUpdated: number = 0;
+  private currentViewBoxRect: Rect = {
+    x: 0,
+    width: 1,
+    y: 0,
+    height: 1,
+  };
+
+  getUpdateIdentifier() {
+    return this.lastUpdated;
+  }
+
+  private updateIdentifier() {
+    this.lastUpdated = Date.now();
+  }
+
+  setXScale(scale: Scale) {
+    this.xScale = scale;
+    this.updateIdentifier();
```

Comment:
Using this identifier to let others (Drawable) know that the Coordinator has changed, is an interesting pattern.

Alternatively, another way is to have Coordinator keep track of 'listeners', and notify them when specific things (scale, viewBoxRect, etc) change.  e.g.
```ts
// coordinator
enum EventType {
  ViewCoordinatesChanged,
}

addListener(eventType: EventType, callback: Function) {}
dispatchEvent(eventType: EventType) {}

setXScale() {
  this.dispatchEvent(EventType.ViewCoordinatesChanged);
}


// drawable
this.coordinator.addListener(ViewCoordinatesChanged, () => {
  this.isDirty = true;
});

setData(data) {
  this.isDirty = true;              // no need for clearIdentifier() complexity!
  this.rawSeriesData = data;
}

```

With events, it is
- easy to reason about who cares when the Coordinator changes
- easy to add new things that listen for Coordinator updates, instead of comparing identifiers
- easy to add more granular events later; if we wanted to distinguish XScale changes from YScale changes, Events.XScaleChanged + Events.YScaleChanged is simpler than making all listeners do `xScaleIdentifier !== prevXScaleIdentifier` + `yScaleIdentifier !== prevYScaleIdentifier`

wdyt?

Reply:
While interesting concept (responding in order of the original bullets):

1. I can kind of see that but if it is imperative that the flow is clearly visible given that the method is only used in DataDrawable.
2. This is the part that I do not see much benefit of. It is important to do the least amount of work and do heavy work only when required. I do not think knowing precise moment of the coordinator change is important as only thing that you need to know is whether some related state needs to be updated. 
3. Not needed; when scale changes, for instance, the coordinate wholely have to be recomputed.

Extra: It is a bit more cumbersome to removeListener and manage this state than managing one number. 

If it is not too much trouble, I would like to move forward as is for above reasons.

Reply:
Sure thing. If we need to communicate more from coordinator to a renderer (and others), we will/should definitely use an event based programming.